File: dsa/list/list.cpp
#include <iostream>

using namespace std;

template<class T>
class Node{
	public:
		T value;
		Node<T>* next;
		Node(){}
		Node(T _value, Node<T>* _next){
			value = _value;
			next = _next;
		}
};

template<class T>
class List{
	private:
		Node<T>* head = NULL;
	public:
		List(){
			head = new Node<T>();
		}
		void insert(T value){
			Node<T>* t = head;
			while(t->next!=NULL){
				t = t->next;
			}
			t->next = new Node<T>(value, NULL);
		}

		bool find(T value){
			Node<T>* t = head;
			while(t->next!=NULL){
				if(t->next->value==value) return true;
				t = t->next;
			}
			return false;
		}

		bool update(T oldValue, T newValue){
			Node<T>* t = head;
			while(t->next!=NULL){
				if(t->next->value==oldValue){
					t->next->value = newValue;
					return true;
				}
			}
			return false;
		}

		bool remove(T value){
			Node<T> *t = head, *prev;
			while(t->next!=NULL){
				prev = t->next;
				if(t->next->value==value){
					t->next = t->next->next;
					delete prev;
					return true;
				}
				t = t->next;
			}
			return false;
		}

		void display(){
			Node<T>* t = head;
			while(t->next!=NULL){
				cout<<t->next->value<<" -> ";
				t = t->next;
			}
			cout<<"end"<<endl;
		}
};

int main(){
	//List of integers
	List<int> list;
	list.display();
	for(int i=0;i<5;i++){
		list.insert(i);
		list.display();
	}
	for(int i=4;i>=0;i--){
		list.remove(i);
		list.display();
	}
	//List of characters
	List<char> slist;
	for(char a='a';a<='z';a++){
		slist.insert(a);
		slist.display();
	}
	for(char z='z';z>='a';z--){
		slist.remove(z);
		slist.display();
	}
	return 0;
}

File: dsa/list/List.java
public class List{

	private class Node{
		private Object obj;
		private Node next;

		public Node(){
			obj = null;
			next = null;
		}

		public Node(Object obj, Node Next){
			this.obj = obj;
			this.next = next;
		}

		public Node getNext(){
			return next;
		}

		public Object getObject(){
			return obj;
		}

		public void setObject(Object obj){
			this.obj = obj;
		}

		public void setNext(Node next){
			this.next = next;
		}

		public String toString(){
			return "("+obj+","+next+")";
		}
	}
	
	private Node head;
	int itemCount;

	List(){
		head = new Node();;
		itemCount = 0;
	}

	void add(Object obj){
		Node t = head;
		while(t.getNext()!=null){
			t = t.getNext();
		}
		t.setNext(new Node(obj,null));
		itemCount++;
	}

	boolean has(Object obj){
		Node t = head;
		while(t.getNext()!=null){
			t = t.getNext();
			if(t.getObject()==obj) return true;
		}
		return false;
	}

	public Object get(int index){
		Node t = head;
		int i = 0;
		while(t.getNext()!=null){
			t = t.getNext();
			if(i==index) return t.getObject();
			i++;
		}
		return null;
	}

	boolean update(Object oldObj, Object newObj){
		Node t = head;
		while(t.getNext()!=null){
			t = t.getNext();
			if(t.getObject()==oldObj){
				t.setObject(newObj);
				return true;
			}
		}
		return false;
	}

	boolean remove(Object obj){
		Node t = head, prev;
		while(t.getNext()!=null){
			prev = t;
			t = t.getNext();
			if(t.getObject()==obj){
				prev.setNext(t.getNext());
				t = null;
				itemCount--;
				return true;
			}
		}
		return false;
	}

	public String toString(){
		String str = "[";
		Node t = head;
		while(t.getNext()!=null){
			t = t.getNext();
			str += t.getObject();
			if(t.getNext()!=null) str += ", ";
		}
		str += "]";
		return str;
	}

	public static void main(String[] args){
		List list = new List();
		for(int i=1;i<=5;i++){
			System.out.println(list);
			list.add(i);
		}
		System.out.println(list);
		System.out.println("Has 5:"+list.has(5));
		for(int i=5;i>=1;i--){
			list.remove(i);
			System.out.println(list);
		}
		System.out.println("Has 5:"+list.has(5));
	}
}
